### PR TITLE
OpenAI-based rephrasing

### PR DESCRIPTION
**Pull request by liwii**
Implemented `langcheck.augment.rephrase`, which does OpenAI-based paraphrasing of the given prompts.

Here are some examples with prompts picked from https://growthtribe.io/blog/chatgpt-prompts.

```
>>> langcheck.augment.rephrase('Solve this math problem for me and explain any jargon.', openai_args={'engine': 'gpt-35-function-calling'})
['Can you solve this math problem and clarify any technical terminology for me?']
>>> langcheck.augment.rephrase('I’m visiting Japan for two weeks, name three cities I should visit that includes fun, culture and romance.', openai_args={'engine': 'gpt-35-function-calling'})
['Name three cities in Japan that I should visit for two weeks, incorporating fun, culture, and romance.']
```

### CODE REVIEW

File: src/langcheck/augment/en/_rephrase.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
+from __future__ import annotations
+
+from typing import Optional
+
+import openai
+
+
+def rephrase(
+        instances: list[str] | str,
```
Comment (kennysong) [inline src/langcheck/augment/en/_rephrase.py:9]: Could you add a num_perturbations parameter so it's consistent with the other augmentations?

File: src/langcheck/augment/en/_rephrase.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        openai_args: Optional[dict[str, str]] = None) -> list[Optional[str]]:
+    '''Rephrase each string in instances (usually a list of prompts) without
+    changintg the meanings. We use the modified version of a prompt presented
+    in `Rethinking Benchmark and Contamination for Language Models with
+    Rephrased Samples <https://arxiv.org/abs/2311.04850>`__ to make an LLM
+    rephrase the given texts.  Currently only OpenAI-based is supported.
+    We use OpenAI's 'gpt-turbo-3.5' model by default. See
+    `this example <https://langcheck.readthedocs.io/en/latest/metrics.html
+    #computing-metrics-with-openai-models>`__
+    for examples on setting up the OpenAI API key.
```
Comment (kennysong) [inline src/langcheck/augment/en/_rephrase.py:19]:
```suggestion
    '''Rephrases each string in instances (usually a list of prompts) without
    changing their meaning. We use a modified version of the prompt presented
    in `"Rethinking Benchmark and Contamination for Language Models with
    Rephrased Samples" <https://arxiv.org/abs/2311.04850>`__ to make an LLM
    rephrase the given text.

    Currently, this is not available locally and requires an OpenAI API key,
    using the 'gpt-turbo-3.5' model by default. See `this page
    <https://langcheck.readthedocs.io/en/latest/metrics.html#computing-metrics-with-openai-models>`__
    for examples on setting up the OpenAI API key.
```

**Review by kennysong — COMMENTED** [review]
This is cool! I reviewed the code but wasn't able to test this locally due to some `openai` library errors. 

Can you rebase this on top of https://github.com/citadel-ai/langcheck/pull/57?

**Comment by liwii** [comment]
Did the rebase!! Confirmed the code now works with openai >= 1. There are some type errors, which should exist in #57 too, but I will work on that after its functionality is reviewed.

**Comment by kennysong** [comment]
Hmm, I wasn't able to get this working. Am I missing something obvious?

```
>>> os.environ["AZURE_OPENAI_KEY"] = '_________'
>>> os.environ["OPENAI_API_VERSION"] = '0301'
>>> os.environ["AZURE_OPENAI_ENDPOINT"] = 'https://openai-kenny-test.openai.azure.com'
>>> langcheck.augment.rephrase('write me an email inviting sam to my birthday party')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workspaces/langcheck/src/langcheck/augment/en/_rephrase.py", line 50, in rephrase
    openai_client = OpenAI()
  File "/home/vscode/.local/lib/python3.8/site-packages/openai/_client.py", line 93, in __init__
    raise OpenAIError(
openai.OpenAIError: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable
>>> from openai import AzureOpenAI
>>> client = AzureOpenAI(api_key='_________',
...                      api_version='0301',
...                      azure_endpoint='https://openai-kenny-test.openai.azure.com')
>>> langcheck.augment.rephrase('write me an email inviting sam to my birthday party', openai_client=client)
OpenAI failed to return a rephrased prompt: Error code: 404 - {'error': {'code': '404', 'message': 'Resource not found'}}
Prompt that triggered the failure is:

            Please rephrase the following prompt without altering its meaning,
            ensuring you adjust the word order appropriately.
            Ensure that no more than five consecutive words are repeated
            and try to use similar words as substitutes where possible.
            [BEGIN DATA]
            ************
            [Prompt]: write me an email inviting sam to my birthday party
            ************
            [END DATA]
            
[None]
```

**Comment by yosukehigashi** [comment]
> Hmm, I wasn't able to get this working. Am I missing something obvious?
> 
> ```
> >>> os.environ["AZURE_OPENAI_KEY"] = '_________'
> >>> os.environ["OPENAI_API_VERSION"] = '0301'
> >>> os.environ["AZURE_OPENAI_ENDPOINT"] = 'https://openai-kenny-test.openai.azure.com'
> >>> langcheck.augment.rephrase('write me an email inviting sam to my birthday party')

If you change this to `langcheck.augment.rephrase('write me an email inviting sam to my birthday party', model_type='azure_openai')`, I think it should work!

**Comment by kennysong** [comment]
Hmmm I tried that and a few variations but I'm getting the same error:

```
>>> import openai, langcheck, os
>>> os.environ["AZURE_OPENAI_KEY"] = '_____________'
>>> os.environ["OPENAI_API_VERSION"] = '0301'
>>> os.environ["AZURE_OPENAI_ENDPOINT"] = 'https://openai-kenny-test.openai.azure.com'
>>> langcheck.augment.rephrase('write me an email inviting sam to my birthday party', model_type='azure_openai')
OpenAI failed to return a rephrased prompt: Error code: 404 - {'error': {'code': '404', 'message': 'Resource not found'}}
Prompt that triggered the failure is:

            Please rephrase the following prompt without altering its meaning,
            ensuring you adjust the word order appropriately.
            Ensure that no more than five consecutive words are repeated
            and try to use similar words as substitutes where possible.
            [BEGIN DATA]
            ************
            [Prompt]: write me an email inviting sam to my birthday party
            ************
            [END DATA]
            
[None]
>>> langcheck.augment.rephrase('write me an email inviting sam to my birthday party', model_type='azure_openai', openai_args={'model': 'langcheckchat-bot'})
OpenAI failed to return a rephrased prompt: Error code: 404 - {'error': {'code': '404', 'message': 'Resource not found'}}
Prompt that triggered the failure is:

            Please rephrase the following prompt without altering its meaning,
            ensuring you adjust the word order appropriately.
            Ensure that no more than five consecutive words are repeated
            and try to use similar words as substitutes where possible.
            [BEGIN DATA]
            ************
            [Prompt]: write me an email inviting sam to my birthday party
            ************
            [END DATA]
            
[None]
>>> langcheck.augment.rephrase('write me an email inviting sam to my birthday party', model_type='azure_openai', openai_args={'model': 'gpt-35-turbo'})
OpenAI failed to return a rephrased prompt: Error code: 404 - {'error': {'code': '404', 'message': 'Resource not found'}}
Prompt that triggered the failure is:

            Please rephrase the following prompt without altering its meaning,
            ensuring you adjust the word order appropriately.
            Ensure that no more than five consecutive words are repeated
            and try to use similar words as substitutes where possible.
            [BEGIN DATA]
            ************
            [Prompt]: write me an email inviting sam to my birthday party
            ************
            [END DATA]
            
[None]
```

**Comment by yosukehigashi** [comment]
ohhh the OPENAI_API_VERSION should be `'2023-03-15-preview'` and not `'0301'`. So once you change that env var, your second one (`langcheck.augment.rephrase('write me an email inviting sam to my birthday party', model_type='azure_openai', openai_args={'model': 'langcheckchat-bot'})`) should work!

**Comment by kennysong** [comment]
> ohhh the OPENAI_API_VERSION should be `'2023-03-15-preview'` and not `'0301'`.

Ah that worked, thanks! Where do you see this version in the Azure console?

**Review by kennysong — COMMENTED** [review]
Tested and this works pretty well! Left a few minor comments

```
>>> langcheck.augment.rephrase(['write me an email inviting sam to my birthday party', 'summarize this document into 50 words'], openai_client=client, openai_args={'model': 'langcheckchat-bot'}, num_perturbations=3)
['Send an email to Sam asking them to attend my birthday celebration.', "Can you compose an email to request Sam's presence at my birthday celebration?", 'Can you draft an email summoning Sam to my birthday celebration?', 'Can you provide a summary of this document using only 50 words?', 'Please summarize the document in 50 words.', 'In 50 words, provide a summary of this document.']
```

File: src/langcheck/augment/en/_rephrase.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+        if model_type == 'openai':
+            openai_client = OpenAI()
+        elif model_type == 'azure_openai':
+            openai_client = AzureOpenAI(
+                api_key=os.getenv("AZURE_OPENAI_KEY"),
+                api_version=os.getenv("OPENAI_API_VERSION"),
+                azure_endpoint=os.getenv(
+                    "AZURE_OPENAI_ENDPOINT"))  # type: ignore
+        else:
+            raise AssertionError(f'Unexpected model type "{model_type}"')
```
Comment (kennysong) [inline src/langcheck/augment/en/_rephrase.py:58]: Should we add this error message like we have for OpenAI metrics?

```
        if self._client_type == 'azure_openai' and not openai_args:
            raise AssertionError(
                'The model deployment must be specified in `openai_args` for '
                'the azure_openai type, e.g. '
                '`openai_args={"model": "YOUR_DEPLOYMENT_NAME"}`')
```

**Comment by yosukehigashi** [comment]
> > ohhh the OPENAI_API_VERSION should be `'2023-03-15-preview'` and not `'0301'`.
> 
> Ah that worked, thanks! Where do you see this version in the Azure console?

Oh it's not tied to a model deployment, we just need to pick a version that has support for the methods that we call. For chat completions, the oldest one that supports it is `'2023-03-15-preview'` ([ref](https://learn.microsoft.com/en-US/azure/ai-services/openai/reference#chat-completions)) so we often just use that (because this is what we initially used in the hackathon) but you can use a newer one and it should still work.

File: src/langcheck/augment/en/_rephrase.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                if openai_args is None:
+                    response = openai_client.chat.completions.create(
+                        model="gpt-3.5-turbo", messages=messages)
+                else:
+                    response = openai_client.chat.completions.create(
+                        messages=messages,
+                        **openai_args,
+                    )
+                # This metrics-with-openai-models>`__ is necessary to pass pyright
+                # since the openai library is not typed.
```
[the remark this answers is not in the document]
Comment (liwii) [reply src/langcheck/augment/en/_rephrase.py:86]: It was a comment on the type check existed before. Let me delete the comments.

File: src/langcheck/augment/en/_rephrase.py (hunk 3 above)
Comment (liwii) [reply src/langcheck/augment/en/_rephrase.py:65]: Sounds good, added!!

**Comment by liwii** [comment]
Alright fixed!!
@yosukehigashi Is it okay to merge this branch into `upgrade-oai`?

**Comment by yosukehigashi** [comment]
> Alright fixed!! @yosukehigashi Is it okay to merge this branch into `upgrade-oai`?

Yeah sure! (Once Kenny approves it 😄)

File: src/langcheck/augment/en/_rephrase.py [hunk 5]
```diff
@@ -39,16 +40,22 @@ def rephrase(
         openai_client: OpenAI or AzureOpenAI client, default None. If this is
             None, we will attempt to create a default client.
         openai_args: Dict of additional args to pass in to the
-            `openai.ChatCompletion.create` function, default None
+            `client.chat.completions.create` function, default None
```
Comment (kennysong) [inline src/langcheck/augment/en/_rephrase.py:43]:
```suggestion
            ``client.chat.completions.create`` function, default None
```

File: src/langcheck/augment/en/_rephrase.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+    1. The 'openai' type, where we use OpenAI's 'gpt-turbo-3.5' model
+    by default. See `this page
+    <https://langcheck.readthedocs.io/en/latest/metrics.html#computing-metrics-with-openai-models>`__
+    for examples on setting up the OpenAI API key.
+
+    2. The 'azure_openai' type. Essentially the same as the 'openai' type,
+    except that it uses the AzureOpenAI client. Note that you must specify the
+    model to use in `openai_args`, e.g.
+    `openai_args={'model': 'YOUR_DEPLOYMENT_NAME'}`
```
Comment (kennysong) [inline src/langcheck/augment/en/_rephrase.py:32]: Single backticks get rendered as italic, and double backticks as inline code blocks in the documentation. We fixed this on Yosuke's PR too

```suggestion
    model to use in ``openai_args``, e.g.
    ``openai_args={'model': 'YOUR_DEPLOYMENT_NAME'}``
```

**Review by kennysong — APPROVED** [review]
Thanks, LGTM after resolving the last comment!

**Pull request merged**